Circom wrapper's helper function

yugonsan opened the pull request:
Hello, I conducted the issue #16 (Circom Wrapper). It would be best to write `frontend_trait` at first, but I made this helper function in advance. 

## What?
`frontend/circom/mod.rs` is the main code in this PR. The file works for below;
1. Extract R1CS constraints from Circom's `.r1cs` files.
2. Convert the BigInt constraints to Bn254's ScalarField constraints.
3. Convert R1CS constraints suited for the Bn254 pairing curve to a format suitable for `folding-schemes`'s R1CS.
4. Calculate the witness from Circom's `.wasm` file and inputs.
5. Convert R1CS constraints and witness from the Circom format to the `folding-schemes`'s `R1CS` and `z` format.

## Changes

- Added `convert_constraints_bigint_to_scalar` function that maps BigInt constraints to Bn254's ScalarField constraints.
- Implemented `extract_constraints_from_r1cs` to parse constraints from Circom's `.r1cs` files.
- Added `convert_to_folding_schemes_r1cs` to adapt the constraints for `folding-schemes`'s R1CS.
- Implemented `calculate_witness` for witness computation using Circom's `.wasm` file.
- Added `circom_to_folding_schemes_r1cs_and_z` to convert from Circom's R1CS and witness to `folding-schemes` formats.

## How?
In #16, I said that I would use `nova-scotia`, but [ark-circom(circom-compat)](https://github.com/arkworks-rs/circom-compat) is based on arkworks so it is more compatible to `folding-schemes`. <br>So I adopted the `ark-circom` instead of `nova-scotia`, and copy&pasted their codes in `circom/r1cs_reader.rs` and `circom/witness/circom.rs, memory.rs, mod.rs, witness_calculator.rs` because I couldn't import `ark-circom` crate due to arkworks dependencies problem. <br>`ark-circom` has Apache-2.0 and MIT license, but it would be better to write credit for the library in readme.

## Testing?

The `test_circom_to_folding_conversion` function showcases the entire process: from extracting constraints from an R1CS file, through conversion to `folding-schemes`'s R1CS and z, to checking the equality of A•z◎B•z=C•z.
I confirmed `folding-schemes`'s R1CS and z from Circom's `.r1cs` and `witness.wasm` satisfy the `check_relation` function in `src/ccs`.

## Anything Else?
### To Do
- Currently, in the convert_to_folding_schemes_r1cs function, the variable l in R1CS is derived from the length of the first row in the a_matrix:
`let l = a_matrix.first().map(|vec| vec.len()).unwrap_or(0);`
This does not accurately represent the size of public inputs and outputs. In other words, everything is public i/o. Moving forward, we need to modify this to ensure l captures the correct size of public inputs and outputs."
- design the `frontend_trait` and fit this circom wrapper function to it.
- mention the credit to ark-circom in readme.

Thank you for your review. I appreciate it and look forward to your feedback!

arnaucube commented:
> So I adopted the `ark-circom` instead of `nova-scotia`, and copy&pasted their codes in `circom/r1cs_reader.rs` and `circom/witness/circom.rs, memory.rs, mod.rs, witness_calculator.rs` because I couldn't import `ark-circom` crate due to arkworks dependencies problem.

Which were the dependencies problems that you found with https://github.com/arkworks-rs/circom-compat ?
I'm afraid that copy pasting part of it's code into the `folding-schemes` repo will bring extra mainteinance overhead, while `circom-compat`/`ark-circom` seems to be well maintained.
Ideally we would use its methods by importing it as a dependency, that's why I'm curious on which issues you found when trying to use it as a dependency.

yugonsan commented:
Successfully imported `ark-circom` as an external library and removed duplicated code. Please review the changes.

arnaucube commented on Cargo.toml:
```diff
@@ -4,22 +4,25 @@ version = "0.1.0"
 ark-relations = { version = "^0.4.0", default-features = false }
 ark-r1cs-std = { version = "^0.4.0", default-features = false }
+ark-circom = { git = "https://github.com/gakonst/ark-circom.git" }
+ark-bn254 = "0.4.0"
```
One question, do you know if ark-circom / circom-compat can work with other curves (with arkworks generics) rather than just bn254? if so (ideally but I'm not sure if it's possible) we could avoid importing bn254 as dependency (but again I'm not sure if ark-circom supports other curves)

yugonsan replied:
Sorry for the late reply. It took me a while to look into it.

I think ark-circom only directly supports `bn254`, but I noticed that it is also possible with other elliptic curves. ark-circom takes `ark_ec::pairing::Pairing;` as generics in R1CS. The R1CS constraints below are the ones with 3 (A,B,C) sparse matrices and depend on `E(pairing)`. Since BN254 implemented the `pairing trait`, ark-circom runs with BN254.

If other elliptic curves also implement the pairing trait, then ark-circom can be run on other elliptic curves as well. Actually, `ark_ec::models::bls12::Bls12`, a dependency of `ark_bls12_381,` has a `pairing trait`, so it should be possible on other elliptic curves would also be possible.
```
pub struct R1CS<E: Pairing> {
    pub num_inputs: usize,
    pub num_aux: usize,
    pub num_variables: usize,
    pub constraints: Vec<Constraints<E>>,
    pub wire_mapping: Option<Vec<usize>>,
}
```
However, the main topic is whether the process of converting from circom to R1CS of folding-schemes can be done in abstract generics, omitting `ark-bn254`. The above R1CS constraints are defined by a `<E as Pairing>::ScalarField` for the value of the sparse matrix, and the `ScalarField` implements a `PrimeField trait`. 
This suggests that it is possible to output R1CS on `PrimeField` of folding-schemes from R1CS of circom, which is easier to handle from the `frontend trait` if unified with `PrimeField`.
```
pub type Constraints<E> = (ConstraintVec<E>, ConstraintVec<E>, ConstraintVec<E>);
pub type ConstraintVec<E> = Vec<(usize, <E as Pairing>::ScalarField)>;
```

I felt the mechanics of ark-works and rust were complicated to me, but this is what I understand now. I am planning to rewrite the implementation over the weekend.

yugonsan replied:
@arnaucube 
Finally, I could get converted R1CS for `folding_schemes` from `ark-circom` without the `BN254` curve. Actually, the `BN254` remains to be used in the test function, because the `ark-cricom`'s `Pairing` generic needs specific curves to try test functions. However, by removing the `BN254` dependencies in the main code, I think the `frontend trait` can easily deal with getting `folding_schemes`'s `R1CS` which is on `PrimeField` rather than `BN254`.

I would be very happy to review these codes again. If possible, I would like to work on the design of the `frontend trait` next. Sorry for the back and forth on the order of issues.

Thank you.

arnaucube replied:
Oh that's very cool!!
Yeah for the tests is fine, we also use specific curves for some tests in other directories only for tests.

Maybe the dependency now can be moved into Cargo.toml's line 26 under `[dev-dependencies]`?

yugonsan replied:
I moved `ark-bn254` into `[dev-dependencies]`.

arnaucube commented on src/frontend/circom/mod.rs:
```diff
@@ -0,0 +1,182 @@
+        let inputs = vec![
+            // success
+            ("step_in".to_string(), vec![BigInt::from(3)]),
+            // fail
+            // ("step_in".to_string(), vec![BigInt::from(6)]),
```
How do you see having a test or part of the test for this value instead of having the test that expects it to fail commented? in this way that part of the test ensures that it fails as expected.

yugonsan replied:
Thank you for the feedback! I changed the test functions have been segregated into success and failure.

arnaucube reviewed (commented):
Very nice! Thank you very much for contribuiting ^^

For task #16 there is a dependancy on #15, but while we don't have #15 (`frontend` trait) defined, this PR seems a good advancement of the methods that will be used once the `frontend` trait is defined.

I don't have much time now (currently OOO), but leaving some comments as a first high level pass.

arnaucube commented on src/frontend/circom/mod.rs:
```diff
@@ -0,0 +1,211 @@
+        let inputs = vec![("step_in".to_string(), vec![BigInt::from(3)])];
+        let witness = calculate_witness(&wasm_filepath, inputs).expect("Error");
+        assert!(!witness.is_empty());
+
+        let (r1cs, z) = circom_to_folding_schemes_r1cs_and_z(converted_constraints, &witness, pub_io_len)
+            .expect("Error");
+        assert!(!z.is_empty());
+        r1cs.check_relation(&z).expect("Error");
+    }
+
+    #[test]
+    fn test_circom_conversion_failure() {
+        let current_dir = env::current_dir().unwrap();
+        let base_path = current_dir.join("src/frontend/circom/test_folder");
+        let r1cs_filepath = base_path.join("test_circuit.r1cs");
+        let wasm_filepath = base_path.join("test_circuit.wasm");
+        
+        assert!(r1cs_filepath.exists());
+        assert!(wasm_filepath.exists());
+        
+        let (constraints, pub_io_len) = extract_constraints_from_r1cs::<Bn254, Fr>(&r1cs_filepath).expect("Error");
+        assert!(!constraints.is_empty());
+
+        let converted_constraints: Vec<Constraints<Fr>> = constraints
+            .iter()
+            .map(|constraint| convert_constraints_bigint_to_scalar::<Bn254, Fr>(constraint.clone()))
+            .collect();
+        assert_eq!(constraints.len(), converted_constraints.len());
+
+        let inputs = vec![("step_in".to_string(), vec![BigInt::from(6)])];
+        let witness = calculate_witness(&wasm_filepath, inputs).expect("Error");
+        assert!(!witness.is_empty());
+
+        let (r1cs, z) = circom_to_folding_schemes_r1cs_and_z(converted_constraints, &witness, pub_io_len)
+            .expect("Error");
+        assert!(!z.is_empty());
+        r1cs.check_relation(&z).expect("Error");
+    }
+
+}
```
Since most of the logic of `test_circom_conversion_success` test is repeated in `test_circom_conversion_failure`, could we abstract these tests to reuse the same test logic being called in different instantiations?
Eg. all of them do the same logic which can be in a test function, with 2 parameters being passed: the inputs vector (where one call of the test passes a correct one and the other call passes an invalid one), and a boolean parameter that is passed to indicate if we expect the test to pass or to fail.

For example similar to what is done here https://github.com/microsoft/Nova/blob/200b47edab190c06a44cbd87a63d34a444184a41/src/nifs.rs#L381 to test the same logic but with different instantiations, but in this PR case it could be a test

```rust
#[test]
fn test_circom_conversion() {
  // expect it to pass for correct inputs
  test_circom_conversion_expect(true, vec![("step_in".to_string(), vec![BigInt::from(3)])];);
  // expect it to fail for incorrect inputs
  test_circom_conversion_expect(false, vec![("step_in".to_string(), vec![BigInt::from(6)])];);
}
```

where the function `test_circom_conversion_expect` contains the actual test logic. In this way we avoid having the same code twice.

yugonsan replied:
Thank you for your guidance on writing readable code!
I believe I've addressed your feedback in the current test function.
Additionally, I think I've resolved the error with the GitHub Actions.
Would you mind taking another look at the code?

arnaucube reviewed (approved):
Tagging @han0110 and @CPerezz for a review

arnaucube commented on src/frontend/circom/mod.rs:
```diff
@@ -0,0 +1,229 @@
+type ExtractedConstraintsResult<F> = Result<(Vec<Constraints<F>>, usize), Box<dyn Error>>;
+
+// Convert the generic type of constraints from Pairing's ScalarField to PrimeField.
+pub fn convert_constraints_bigint_to_scalar<E, F>(
+    constraints: Constraints<E::ScalarField>,
+) -> Constraints<F>
+where
+    E: Pairing,
+    F: PrimeField,
+{
+    let convert_vec = |vec: ConstraintVec<E::ScalarField>| -> ConstraintVec<F> {
+        vec.into_iter()
+            .map(|(index, element)| {
+                // Convert element into BigInt, then to BigUint, then to the destination PrimeField
+                let bigint_e: <<E as Pairing>::ScalarField as PrimeField>::BigInt =
+                    element.into_bigint();
+                let generic_biguint: num_bigint::BigUint = bigint_e.into();
+                let f_element: F = F::from(generic_biguint);
```
Do you think that we can get rid of the `F` since would be the same as `E::ScalarField`?

han0110 replied:
I also feel this function seems unnecessary, I think we could have `extract_constraints_from_r1cs` to be like:

```rust
pub fn extract_constraints_from_r1cs<E>(
    r1cs_filepath: &Path,
) -> ExtractedConstraintsResult<E::ScalarField>
where
    E: Pairing,
{
    // ...
}
```

CPerezz replied:
I don't like this either.
We have one function which body is one closure.
Couldn't we do it on another way?

I think @han0110 's suggestion looks good though

yugonsan replied:
Thank you for the feedback. I've removed the convert_constraints_bigint_to_scalar function as it indeed was completely unnecessary. The code is now simplified by using only E::ScalarField.

han0110 commented on src/frontend/circom/mod.rs:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    let l = pub_io_len;
+    let n_cols = a_matrix.first().map(|vec| vec.len()).unwrap_or(0);
```
This seems incorrect to me, it should be `r1cs.num_variables` iiuc.

yugonsan replied:
Thank you for your feedback. I flexed this problem.

CPerezz reviewed (commented):
Could we avoid adding `.wasm` and `.r1cs` files for testing and generate them on the fly? Would that take a lot of time and slow down the tests that much? 

Asking as otherwise, sync the repo will take time. And if we can save it would be cool. But not at the cost of tests taking forever. 
In any case, worth exploring the alternatives.


Also, as for the frontend, could we have some struct/trait for which we implement this public functions? I'd rather have a unit struct that has nothing but parses all that stuff so that I don't need to call 3 functions rather than this.

yugonsan commented:
@han0110 Thanks for the feedback.
> Could we avoid adding .wasm and .r1cs files for testing and generate them on the fly? Would that take a lot of time and slow down the tests that much?

Certainly, I believe it's possible by using a sh file. In fact, `nova-scotia` allows us to generate `.wasm` and `.r1cs` files from `.circom` directly within a sh file, as seen [(here)](https://github.com/nalinbhardwaj/Nova-Scotia/blob/main/examples/toy/compile.sh). I'll work on implementing this shortly.

> Also, as for the frontend, could we have some struct/trait for which we implement this public functions? I'd rather have a unit struct that has nothing but parses all that stuff so that I don't need to call 3 functions rather than this.

Thank you for the suggestion. I see the merit in your approach and will proceed with creating such a struct or trait.

yugonsan commented:
I've addressed and refined every point in our code based on your invaluable feedback. Thank you for your guidance. Here's a concise summary of the modifications:

I removed the redundant generic `F`, now using `E::ScalarField directly` in the proper functions, making the code cleaner. And then, the unnecessary `convert_constraints_bigint_to_scalar` function has been removed. 

Instead of adding `.wasm` and `.r1cs` files directly, we now generate them on the fly through the `sh` file during tests, ensuring a lightweight repository.

I integrated a `CircomFrontend` trait within the `CircomWrapper` struct, offering a more streamlined interface.

I would be most grateful if you could review the revised code again. @arnaucube

han0110 reviewed (approved):
LGTM

CPerezz reviewed (approved):
Fixed conflicts.

LGTM! Thanks for this work @yugonsan

yugonsan commented on src/frontend/circom/test_folder/compile.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+circom ./src/frontend/circom/test_folder/test_circuit.circom --r1cs --wasm --prime bn128 --output ./src/frontend/circom/test_folder/
```
I think I improved this issue. I'd like to see the GitHub action's result.

yugonsan commented:
I have fixed some issues based on the error in GitHub actions. I have edited the yaml file to install `Circom` in GitHub Actions, and I would be happy to run GitHub Actions again.

arnaucube commented on .github/workflows/ci.yml:
```diff
@@ -49,6 +49,14 @@ jobs:
       # use the more efficient nextest
       - uses: taiki-e/install-action@nextest
       - uses: Swatinem/rust-cache@v2
+      - name: Install Circom
+        run: |
+          git clone https://github.com/iden3/circom.git
+          cd circom
+          cargo build --release
+          cargo install --path circom
+      - name: Execute compile.sh
+        run: bash ./src/frontend/circom/test_folder/compile.sh
```
This will take too many minutes each time, since it downloads and compiles from source code.
Since some months ago, we can directly download the compiled binaries from the releases.
Maybe we can change these lines by:

```suggestion
      - name: Download Circom
        run: |
          mkdir -p $HOME/bin
          curl -sSfL https://github.com/iden3/circom/releases/download/v2.1.6/circom-linux-amd64 -o $HOME/bin/circom
          chmod +x $HOME/bin/circom
          echo "$HOME/bin" >> $GITHUB_PATH
```

yugonsan replied:
Thank you for telling me. It must be better.

arnaucube reviewed (approved):
Very good job! The interface looks very nice now^^
I'm excited to start folding Circom circuits soon™
